File: acs/cwmp.py
from __future__ import annotations

from typing import Dict, Optional, Tuple
import xml.etree.ElementTree as ET

SOAP_ENV_URI = "http://schemas.xmlsoap.org/soap/envelope/"
CWMP_DEFAULT_URI = "urn:dslforum-org:cwmp-1-2"

# Register common namespaces for pretty output
ET.register_namespace("SOAP-ENV", SOAP_ENV_URI)
ET.register_namespace("cwmp", CWMP_DEFAULT_URI)
ET.register_namespace("xsd", "http://www.w3.org/2001/XMLSchema")
ET.register_namespace("xsi", "http://www.w3.org/2001/XMLSchema-instance")


def get_localname(tag: str) -> str:
    if tag.startswith("{"):
        return tag.split("}", 1)[1]
    return tag


def find_child_by_localname(parent: ET.Element, localname: str) -> Optional[ET.Element]:
    for child in list(parent):
        if get_localname(child.tag) == localname:
            return child
    return None


def parse_request(xml_bytes: bytes) -> Tuple[str, Optional[str], ET.Element, str]:
    """
    Parse an incoming CWMP SOAP request.

    Returns: (method_name, cwmp_id, method_elem, cwmp_namespace_uri)
    - method_name: e.g. "Inform", "TransferComplete"; "" if empty body
    - cwmp_id: the cwmp:ID value if present, else None
    - method_elem: the first element in Body (if present)
    - cwmp_namespace_uri: CWMP namespace URI to reuse in responses
    """
    root = ET.fromstring(xml_bytes)

    header = find_child_by_localname(root, "Header")
    body = find_child_by_localname(root, "Body")
    if body is None:
        raise ValueError("SOAP Body not found")

    cwmp_id = None
    cwmp_ns_uri = CWMP_DEFAULT_URI

    if header is not None:
        id_elem = None
        for child in list(header):
            if get_localname(child.tag) == "ID":
                id_elem = child
                break
        if id_elem is not None:
            cwmp_id = (id_elem.text or "").strip() or None
            if id_elem.tag.startswith("{"):
                cwmp_ns_uri = id_elem.tag[1:].split("}", 1)[0] or cwmp_ns_uri

    method_elem = None
    for child in list(body):
        # skip text/whitespace nodes; ElementTree returns only elements here
        method_elem = child
        break

    if method_elem is None:
        return "", cwmp_id, body, cwmp_ns_uri

    if method_elem.tag.startswith("{"):
        cwmp_ns_uri = method_elem.tag[1:].split("}", 1)[0] or cwmp_ns_uri

    method_name = get_localname(method_elem.tag)
    return method_name, cwmp_id, method_elem, cwmp_ns_uri


def build_envelope(cwmp_ns_uri: str, cwmp_id: Optional[str]) -> Tuple[ET.Element, ET.Element]:
    """Create a SOAP Envelope with Header (echo cwmp:ID if provided) and Body."""
    # Ensure response uses request CWMP namespace for maximum compatibility
    ET.register_namespace("cwmp", cwmp_ns_uri)

    env = ET.Element(f"{{{SOAP_ENV_URI}}}Envelope")

    header = ET.SubElement(env, f"{{{SOAP_ENV_URI}}}Header")
    if cwmp_id:
        id_elem = ET.SubElement(header, f"{{{cwmp_ns_uri}}}ID")
        id_elem.set(f"{{{SOAP_ENV_URI}}}mustUnderstand", "1")
        id_elem.text = cwmp_id

    body = ET.SubElement(env, f"{{{SOAP_ENV_URI}}}Body")
    return env, body


def build_inform_response(cwmp_id: Optional[str], cwmp_ns_uri: str, max_envelopes: int = 1) -> bytes:
    env, body = build_envelope(cwmp_ns_uri, cwmp_id)
    resp = ET.SubElement(body, f"{{{cwmp_ns_uri}}}InformResponse")
    max_env = ET.SubElement(resp, "MaxEnvelopes")
    max_env.text = str(max_envelopes)
    return serialize_xml(env)


def build_transfer_complete_response(cwmp_id: Optional[str], cwmp_ns_uri: str) -> bytes:
    env, body = build_envelope(cwmp_ns_uri, cwmp_id)
    ET.SubElement(body, f"{{{cwmp_ns_uri}}}TransferCompleteResponse")
    return serialize_xml(env)


def parse_inform(method_elem: ET.Element) -> Dict:
    """Extract useful data from an Inform RPC for storage."""
    device_id_data = {
        "manufacturer": "",
        "oui": "",
        "productClass": "",
        "serialNumber": "",
    }

    device_id_elem = find_child_by_localname(method_elem, "DeviceId")
    if device_id_elem is not None:
        man = find_child_by_localname(device_id_elem, "Manufacturer")
        oui = find_child_by_localname(device_id_elem, "OUI")
        pclass = find_child_by_localname(device_id_elem, "ProductClass")
        sn = find_child_by_localname(device_id_elem, "SerialNumber")
        device_id_data["manufacturer"] = (man.text or "").strip() if man is not None else ""
        device_id_data["oui"] = (oui.text or "").strip() if oui is not None else ""
        device_id_data["productClass"] = (pclass.text or "").strip() if pclass is not None else ""
        device_id_data["serialNumber"] = (sn.text or "").strip() if sn is not None else ""

    events: list[Dict[str, str]] = []
    event_list_elem = find_child_by_localname(method_elem, "Event")
    if event_list_elem is not None:
        for ev in list(event_list_elem):
            if get_localname(ev.tag) != "EventStruct":
                continue
            code = find_child_by_localname(ev, "EventCode")
            cmdk = find_child_by_localname(ev, "CommandKey")
            events.append(
                {
                    "eventCode": (code.text or "").strip() if code is not None else "",
                    "commandKey": (cmdk.text or "").strip() if cmdk is not None else "",
                }
            )

    params: Dict[str, str] = {}
    plist = find_child_by_localname(method_elem, "ParameterList")
    if plist is not None:
        for p in list(plist):
            if get_localname(p.tag) != "ParameterValueStruct":
                continue
            name_elem = find_child_by_localname(p, "Name")
            value_elem = find_child_by_localname(p, "Value")
            if name_elem is None or value_elem is None:
                continue
            name = (name_elem.text or "").strip()
            value = (value_elem.text or "").strip()
            if name:
                params[name] = value

    return {
        "deviceId": device_id_data,
        "events": events,
        "parameterValues": params,
    }


def serialize_xml(elem: ET.Element) -> bytes:
    xml_bytes = ET.tostring(elem, encoding="utf-8")
    return b"<?xml version=\"1.0\" encoding=\"utf-8\"?>" + xml_bytes

File: src/cwmp.js
const { XMLParser, XMLBuilder } = require('fast-xml-parser');
const { nanoid } = require('nanoid');

const xmlParser = new XMLParser({
  ignoreAttributes: false,
  attributeNamePrefix: '@_',
  removeNSPrefix: false,
  allowBooleanAttributes: true,
  parseTagValue: true,
  parseAttributeValue: true,
});

const xmlBuilder = new XMLBuilder({
  ignoreAttributes: false,
  attributeNamePrefix: '@_',
  suppressEmptyNode: true,
  format: true,
});

function parseSoap(xmlString) {
  try {
    const json = xmlParser.parse(xmlString);
    return json;
  } catch (e) {
    const error = new Error('Failed to parse SOAP XML');
    error.cause = e;
    throw error;
  }
}

function getCwmpMethodName(envelope) {
  const body = envelope?.['s:Envelope']?.['s:Body'] || envelope?.Envelope?.Body || envelope?.['soap-env:Envelope']?.['soap-env:Body'];
  if (!body) return null;
  const keys = Object.keys(body);
  if (keys.length === 0) return null;
  // Skip Fault and headers
  const methodKey = keys.find((k) => !['Fault'].includes(k));
  return methodKey || null;
}

function buildSoapEnvelope({ header = {}, body }) {
  const envelope = {
    'soap-env:Envelope': {
      '@_xmlns:soap-env': 'http://schemas.xmlsoap.org/soap/envelope/',
      '@_xmlns:soap-enc': 'http://schemas.xmlsoap.org/soap/encoding/',
      '@_xmlns:xsd': 'http://www.w3.org/2001/XMLSchema',
      '@_xmlns:xsi': 'http://www.w3.org/2001/XMLSchema-instance',
      '@_xmlns:cwmp': 'urn:dslforum-org:cwmp-1-0',
      'soap-env:Header': header,
      'soap-env:Body': body,
    },
  };
  return xmlBuilder.build(envelope);
}

function buildInformResponse({ maxEnvelopes = 1, idHeader }) {
  const header = {};
  if (idHeader) {
    header['cwmp:ID'] = { '@_soap-env:mustUnderstand': 1, '#text': idHeader };
  }
  const body = {
    'cwmp:InformResponse': {
      'MaxEnvelopes': maxEnvelopes,
    },
  };
  return buildSoapEnvelope({ header, body });
}

function extractInformData(envelope) {
  const body = envelope?.['s:Envelope']?.['s:Body']
    || envelope?.Envelope?.Body
    || envelope?.['soap-env:Envelope']?.['soap-env:Body'];
  if (!body) return null;
  const inform = body['cwmp:Inform'] || body.Inform || body['cwmp-1-0:Inform'];
  if (!inform) return null;
  const deviceId = inform?.DeviceId || inform?.DeviceID;
  const oui = deviceId?.OUI || deviceId?.Oui || deviceId?.oui || null;
  const serialNumber = deviceId?.SerialNumber || deviceId?.Serialnumber || null;
  const productClass = deviceId?.ProductClass || deviceId?.Productclass || null;
  const manufacturer = deviceId?.Manufacturer || null;

  // ConnectionRequestURL often inside ParameterList items
  let connectionRequestURL = null;
  const parameterList = inform?.ParameterList?.ParameterValueStruct || inform?.ParameterList || [];
  const paramsArray = Array.isArray(parameterList) ? parameterList : [parameterList];
  for (const p of paramsArray) {
    const name = p?.Name || p?.name;
    const value = p?.Value || p?.Value?.['#text'] || p?.value;
    if (typeof name === 'string' && name.endsWith('ManagementServer.ConnectionRequestURL')) {
      connectionRequestURL = typeof value === 'object' ? value?.['#text'] : value;
    }
  }

  return {
    manufacturer,
    oui,
    productClass,
    serialNumber,
    connectionRequestURL,
  };
}

function buildGetParameterValues({ parameterNames, id }) {
  const idHeader = id || nanoid();
  const header = {
    'cwmp:ID': { '@_soap-env:mustUnderstand': 1, '#text': idHeader },
  };
  const namesArray = Array.isArray(parameterNames) ? parameterNames : [parameterNames];
  const body = {
    'cwmp:GetParameterValues': {
      'ParameterNames': {
        '@_soap-enc:arrayType': `xsd:string[${namesArray.length}]`,
        'string': namesArray,
      },
    },
  };
  return { xml: buildSoapEnvelope({ header, body }), id: idHeader };
}

function buildReboot({ commandKey = '', id }) {
  const idHeader = id || nanoid();
  const header = {
    'cwmp:ID': { '@_soap-env:mustUnderstand': 1, '#text': idHeader },
  };
  const body = {
    'cwmp:Reboot': {
      'CommandKey': commandKey,
    },
  };
  return { xml: buildSoapEnvelope({ header, body }), id: idHeader };
}

module.exports = {
  parseSoap,
  getCwmpMethodName,
  buildInformResponse,
  extractInformData,
  buildGetParameterValues,
  buildReboot,
};
